direct/core/src/main/java/cz/o2/proxima/direct/transaction/TransactionalOnlineAttributeWriter.java

```diff
@@ -0,0 +1,215 @@
+            "Updated sequence ID. That is a bug in proxima's transactions.");
+        sequenceId = response.getSeqId();
+      }
+      state = State.Flags.OPEN;
+    }
+
+    public void commitWrite(List<StreamElement> outputs, CommitCallback callback)
+        throws TransactionRejectedException {
+
+      List<StreamElement> toWrite =
+          outputs.stream().map(this::injectSequenceId).collect(Collectors.toList());
+      List<KeyAttribute> keyAttributes =
+          toWrite.stream().map(KeyAttributes::ofStreamElement).collect(Collectors.toList());
+      manager.commit(transactionId, keyAttributes);
+      Response response = ExceptionUtils.uncheckedFactory(responseQueue::take);
+      if (response.getFlags() != Response.Flags.COMMITTED) {
+        if (response.getFlags() == Response.Flags.ABORTED) {
+          // already aborted
+          needsRollback = false;
+        }
+        throw new TransactionRejectedException(transactionId);
+      }
+      CommitCallback compositeCallback = CommitCallback.afterNumCommits(toWrite.size(), callback);
+      state = State.Flags.COMMITTED;
+      toWrite.forEach(w -> delegate.write(w, compositeCallback));
+    }
+
+    private void enqueueResponse(String responseId, Response response) {
+      ExceptionUtils.unchecked(() -> responseQueue.put(response));
+    }
+
+    private StreamElement injectSequenceId(StreamElement in) {
+
+      Preconditions.checkArgument(!in.isDeleteWildcard(), "Wildcard deletes not yet supported");
+
+      return StreamElement.upsert(
+          in.getEntityDescriptor(),
+          in.getAttributeDescriptor(),
+          sequenceId,
+          in.getKey(),
+          in.getAttribute(),
+          in.getStamp(),
+          in.getValue());
+    }
+
+    @Override
+    public void close() {
+      if (needsRollback && state != State.Flags.COMMITTED) {
+        rollback();
+      }
+    }
+
+    public void rollback() {
+      manager.rollback(transactionId);
+    }
+  }
+
+  private final OnlineAttributeWriter delegate;
+  private final ClientTransactionManager manager;
+
+  private TransactionalOnlineAttributeWriter(
+      DirectDataOperator direct, OnlineAttributeWriter delegate) {
+
+    this.delegate = delegate;
+    this.manager = TransactionResourceManager.of(direct);
+  }
+
+  @Override
+  public URI getUri() {
+    return delegate.getUri();
+  }
+
+  @Override
+  public void close() {
+    delegate.close();
+  }
+
+  @Override
+  public void write(StreamElement data, CommitCallback statusCallback) {
+    // this means start transaction with the single KeyAttribute as input and then
+    // commit that transaction right away
+    try (Transaction t = begin()) {
+      @Nullable
+      String suffix =
+          data.getAttributeDescriptor().isWildcard()
+              ? data.getAttribute()
+                  .substring(data.getAttributeDescriptor().toAttributePrefix().length())
+              : null;
+      KeyAttribute zeroInput =
```

Comment:
zeroInput?

Reply:
The transaction manager needs to know which attribute(s) are going to be part of the transaction. This is necessary so that correct partitioner can be chosen (this is TBD) for parallelism. If the transaction has no input (because this is pure write), we still need to notify the manager about the attribute. It is not input, though. That is why `zeroInput`.